### PR TITLE
MemTile DMA configuration using block write

### PR DESCRIPTION
**Pull request by pvasireddy-amd**
This PR adds MLIR tests to configure MemTile DMAs using TxN block write and Writebd with locks

### CODE REVIEW

File: test/npu-xrt/memtile_dmas/aie.mlir [hunk 1]
```diff
@@ -0,0 +1,42 @@
+module {
+  aie.device(npu1_1col) {
+    memref.global "public" @out_cons : memref<4096xi32>
+    memref.global "public" @out : memref<4096xi32>
+    memref.global "public" @in_cons : memref<4096xi32>
+    memref.global "public" @in : memref<4096xi32>
+    %tile_0_0 = aie.tile(0, 0)
+    %tile_0_1 = aie.tile(0, 1)
+    // %out_cons_prod_lock = aie.lock(%tile_0_0, 2) {init = 0 : i32, sym_name = "out_cons_prod_lock"}
+    // %out_cons_cons_lock = aie.lock(%tile_0_0, 3) {init = 0 : i32, sym_name = "out_cons_cons_lock"}
```
Comment (jgmelber) [inline test/npu-xrt/memtile_dmas/aie.mlir:10]:
```suggestion
```

File: test/npu-xrt/memtile_dmas/aie.mlir [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    %in_cons_buff_0 = aie.buffer(%tile_0_1) {address = 0 : i32, mem_bank = 0 : i32, sym_name = "in_cons_buff_0"} : memref<4096xi32> 
+    %in_cons_prod_lock = aie.lock(%tile_0_1, 0) {init = 1 : i32, sym_name = "in_cons_prod_lock"}
+    %in_cons_cons_lock = aie.lock(%tile_0_1, 1) {init = 0 : i32, sym_name = "in_cons_cons_lock"}
+    // %in_prod_lock = aie.lock(%tile_0_0, 0) {init = 0 : i32, sym_name = "in_prod_lock"}
+    // %in_cons_lock = aie.lock(%tile_0_0, 1) {init = 0 : i32, sym_name = "in_cons_lock"}
```
Comment (jgmelber) [inline test/npu-xrt/memtile_dmas/aie.mlir:15]:
```suggestion
```

**Review by jgmelber — APPROVED** [review]
You can delete `CMakeLists.txt` and `Makefile`.

File: test/npu-xrt/memtile_dmas/aie.mlir
Comment (fifield) [inline test/npu-xrt/memtile_dmas/aie.mlir:1]: This file needs license and copyright header.

**Review by fifield — APPROVED** [review]
I;m happy to see this working! Is there a `aiex.npu.writebd` version?

File: test/npu-xrt/memtile_dmas/aie.mlir
Comment (pvasireddy-amd) [reply test/npu-xrt/memtile_dmas/aie.mlir:1]: `aiex.npu.writebd` version also works. Added it  to `test/npu-xrt/memtile_dmas/writebd`.

Added license and copyright header.

File: lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp [hunk 3]
```diff
@@ -456,56 +456,99 @@ struct WriteBdToBlockWritePattern : OpConversionPattern<NpuWriteBdOp> {
...
+      // TODO: TLAST Suppress
+      words[7] |= (op.getNextBd() & 0xf) << 27;
+      words[7] |= (op.getUseNextBd() & 0x1) << 26;
+      words[7] |= (op.getValidBd() & 0x1) << 25;
+      words[7] |= (op.getLockRelVal() & 0xef) << 18;
+      words[7] |= (op.getLockRelId() & 0xf) << 13;
+      words[7] |= (op.getLockAcqEnable() & 0x1) << 12;
+      words[7] |= (op.getLockAcqVal() & 0xef) << 5;
+      words[7] |= op.getLockAcqId() & 0xf;
+    } else {
```
Comment (jgmelber) [inline lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp:508]: What about worker tiles? Also some devices might have more than one row of memtiles. Check against the device model

File: lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp [hunk 4]
```diff
@@ -456,56 +456,99 @@ struct WriteBdToBlockWritePattern : OpConversionPattern<NpuWriteBdOp> {
...
-    words[7] |= (op.getValidBd() & 0x1) << 25;
-    words[7] |= (op.getLockRelVal() & 0xef) << 18;
-    words[7] |= (op.getLockRelId() & 0xf) << 13;
-    words[7] |= (op.getLockAcqEnable() & 0x1) << 12;
-    words[7] |= (op.getLockAcqVal() & 0xef) << 5;
-    words[7] |= op.getLockAcqId() & 0xf;
-
+    auto bd_id = op.getBdId();
+    uint32_t bd_addr;
+    if (op.getRow() == 0) {
```
Comment (jgmelber) [inline lib/Dialect/AIEX/Transforms/AIEDmaToNpu.cpp:462]: This should check against the device model for NOC tile

**Comment by fifield** [comment]
This PR exposes an error in `test/Conversion/DmaToNpu/dma_to_npu_width_conversion.mlir`. Try this:
```diff
diff --git a/test/Conversion/DmaToNpu/dma_to_npu_width_conversion.mlir b/test/Conversion/DmaToNpu/dma_to_npu_width_conversion.mlir
index ebcc0b91..a1ca5f9b 100644
--- a/test/Conversion/DmaToNpu/dma_to_npu_width_conversion.mlir
+++ b/test/Conversion/DmaToNpu/dma_to_npu_width_conversion.mlir
@@ -31,7 +31,7 @@ module @shimDmaMemcpy{
       aiex.npu.dma_memcpy_nd (2, 0, %arg0[0, 0, 0, 0][4, 4, 64, 64][0, 64, 256, 1]) {id = 0 : i64, metadata = @toMem} : memref<65536xbf16>
       aiex.npu.sync {channel = 0 : i32, column = 0 : i32, column_num = 1 : i32, direction = 0 : i32, row = 0 : i32, row_num = 1 : i32}
     }
-    aie.shim_dma_allocation @toMem (S2MM, 0, 0)
+    aie.shim_dma_allocation @toMem (S2MM, 0, 2)
   }
 }

File: test/npu-xrt/memtile_dmas/writebd_tokens/aie.mlir [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      aiex.npu.write32 {address = 656900 : ui32, column = 0 : i32, row = 1 : i32, value = 0 : ui32}
+      aiex.npu.writebd {bd_id = 1 : i32, buffer_length = 4096 : i32, buffer_offset = 0 : i32, column = 0 : i32, d0_size = 0 : i32, d0_stride = 0 : i32, d1_size = 0 : i32, d1_stride = 0 : i32, d2_stride = 0 : i32, enable_packet = 0 : i32, iteration_current = 0 : i32, iteration_size = 0 : i32, iteration_stride = 0 : i32, lock_acq_enable = 0 : i32, lock_acq_id = 0 : i32, lock_acq_val = 0 : i32, lock_rel_id = 0 : i32, lock_rel_val = 0 : i32, next_bd = 0 : i32, out_of_order_id = 0 : i32, packet_id = 0 : i32, packet_type = 0 : i32, row = 1 : i32, use_next_bd = 0 : i32, valid_bd = 1 : i32}
+      aiex.npu.write32 {address = 656948 : ui32, column = 0 : i32, row = 1 : i32, value = 1 : ui32}
+      aiex.npu.sync {channel = 0 : i32, column = 0 : i32, column_num = 1 : i32, direction = 0 : i32, row = 1 : i32, row_num = 1 : i32}
```
Comment (fifield) [inline test/npu-xrt/memtile_dmas/writebd_tokens/aie.mlir:30]: I don't think this sync will work (line 30). The write to the task queue on line 29 does not set the enable token issue bit. Even if this is fixed, I am not convinced that the TCT will get routed out to the firmware. At least I couldn't get it to work after fiddling with it for a few minutes.

**Review by fifield — COMMENTED** [review]
I think this is ready to commit if the broken tests using tokens are removed.

File: test/npu-xrt/memtile_dmas/writebd_tokens/aie.mlir (hunk 5 above)
Comment (pvasireddy-amd) [reply test/npu-xrt/memtile_dmas/writebd_tokens/aie.mlir]: Thank you for looking into this! Removed the broken token tests for now.

**Pull request merged**